lib/ballot/ballot_test.go

```diff
@@ -218,3 +219,127 @@ func TestBallotProposerTransaction(t *testing.T) {
 		require.NotNil(t, err)
 	}
 }
+
+func TestNewBallot(t *testing.T) {
+	kp, _ := keypair.Random()
+	nodeEndpoint, _ := common.NewEndpointFromString("https://localhost:1000")
+	proposerEndpoint, _ := common.NewEndpointFromString("https://localhost:1001")
+	n, _ := node.NewLocalNode(kp, nodeEndpoint, "")
+	p, _ := node.NewLocalNode(kp, proposerEndpoint, "")
+
+	round := round.Round{Number: 0, BlockHeight: 1, BlockHash: "hahaha", TotalTxs: 1}
+
+	b := NewBallot(n.Address(), p.Address(), round, []string{})
+
+	require.Equal(t, n.Address(), b.Source())
+	require.Equal(t, p.Address(), b.Proposer())
+}
+
+// In this test, we can check that the normal ballot(not expired) should be signed by proposer.
+func TestIsBallotWellFormed(t *testing.T) {
+	nodeEndpoint, _ := common.NewEndpointFromString("https://localhost:1000")
+	proposerEndpoint, _ := common.NewEndpointFromString("https://localhost:1001")
+
+	nodeKP, _ := keypair.Random()
+	n, _ := node.NewLocalNode(nodeKP, nodeEndpoint, "")
+
+	proposerKP, _ := keypair.Random()
+	p, _ := node.NewLocalNode(proposerKP, proposerEndpoint, "")
+
+	round := round.Round{Number: 0, BlockHeight: 1, BlockHash: "hahaha", TotalTxs: 1}
+
+	initialBalance := common.Amount(common.BaseReserve)
+	kpNewAccount, _ := keypair.Random()
+	tx := transaction.MakeTransactionCreateAccount(nodeKP, kpNewAccount.Address(), initialBalance)
+
+	wellBallot := NewBallot(n.Address(), p.Address(), round, []string{tx.GetHash()})
+
+	commonKP, _ := keypair.Random()
+	commonAccount := block.NewBlockAccount(commonKP.Address(), 0)
+
+	opi, _ := NewInflationFromBallot(*wellBallot, commonAccount.Address, initialBalance)
+	opc, _ := NewCollectTxFeeFromBallot(*wellBallot, commonAccount.Address, tx)
+	ptx, _ := NewProposerTransactionFromBallot(*wellBallot, opc, opi)
+	wellBallot.SetProposerTransaction(ptx)
+
+	wellBallot.Sign(proposerKP, networkID)
+
+	err := wellBallot.IsWellFormed(networkID, common.NewConfig())
+
+	require.Nil(t, err)
+
+	wrongSignedBallot := NewBallot(n.Address(), p.Address(), round, []string{tx.GetHash()})
+	wrongSignedBallot.SetProposerTransaction(ptx)
+
+	wrongSignedBallot.Sign(nodeKP, networkID)
+
+	err = wrongSignedBallot.IsWellFormed(networkID, common.NewConfig())
+
+	require.NotNil(t, err)
+
+}
+
+// We can check that expired ballot could be signed by node key pair(not proposer).
+func TestIsExpiredBallotWellFormed(t *testing.T) {
```

Comment:
@Charleslee522 I think this test, `TestIsExpiredBallotWellFormed` missed something. Without `ProposerTransaction` it must be not-wellformed, I think.

Reply:
> @Charleslee522 I think this test, `TestIsExpiredBallotWellFormed` missed something. Without `ProposerTransaction` it must be not-wellformed, I think.

@spikeekips Oh,, then even `expired` ballot should have `ProposerTransaction`?
I think it's a waste of resources. `expired` ballot is only for checking the node is expired.
So `ProposerTransaction` in expired ballot means nothing.

Reply:
@Charleslee522 Yes, I also think so, it will be matter of choice. After expired ballot is finished, the new block will be stored with block inflation, I wish.

Reply:
@spikeekips 
You are right. In ISAAC, A node cannot reach to ALL-CONFIRM with expired ballot as you know.
If the n of expired ballot is over threshold, the next round starts.
Only `VotingYes` ballots can make the node reach to ALL-CONFIRM.